functionary/train_vision/train.py

```diff
@@ -0,0 +1,539 @@
+        of shape (batch_size x seq_len)"""
+        correct_logits = logits
+        added_labels = None
+        if (
+            type(logits) is tuple
+        ):  # in mixtral logits is a tuple, correct logits is at the second index
+            correct_logits = logits[1]
+        elif type(logits) is dict:
+            correct_logits = logits["logits"]
+            added_labels = logits["labels"]
+            labels = added_labels
+            print(
+                f"shape of labels: {added_labels.shape}, shape of logits: {correct_logits.shape}"
+            )
+
+        pred_ids = torch.argmax(correct_logits, dim=-1)
+
+        loss_fn = CrossEntropyLoss(reduction="none")
+        shift_logits = correct_logits[..., :-1, :].contiguous()
+        shift_labels = labels[..., 1:].contiguous()
+        shift_logits = shift_logits.view(-1, len(tokenizer))
+        shift_labels = shift_labels.view(-1)
+        loss = loss_fn(shift_logits, shift_labels)
+        loss = torch.mean(loss.view(correct_logits.shape[0], -1), dim=-1)
+
+        return pred_ids, loss, added_labels
+
+    def compute_accuracy_metrics(prediction_list, label_list):
+        """Computes next-token accuracy and perplexity metrics for evaluation"""
+        acc_count = 0
+        total_num = 0
+        dic = {token_id: {"acc": 0, "total": 0} for token_id in id2token}
+
+        first_token_total_count, first_token_correct_count = 0, 0
+        first_token_label_dic = {}
+
+        for i in range(len(prediction_list)):
+            pred, label = prediction_list[i], label_list[i]
+            if i > 0 and label_list[i - 1] == -100 and label != -100:  # first token
+                first_token_total_count += 1
+                if label not in first_token_label_dic:
+                    first_token_label_dic[label] = {"correct": 0, "total": 0}
+
+                first_token_label_dic[label]["total"] += 1
+
+                if label == pred:
+                    first_token_correct_count += 1
+                    first_token_label_dic[label]["correct"] += 1
+
+            if label != -100:
+                if label == pred:
+                    acc_count += 1
+                total_num += 1
+            if label in dic:
+                dic[label]["total"] += 1
+                if label == pred:
+                    dic[label]["acc"] += 1
+
+        metrics = {"accuracy": acc_count / total_num}
+        metrics = {
+            "accuracy_first_token": first_token_correct_count / first_token_total_count,
+            "total_number_first_token": first_token_total_count,
+        }
+
+        for token_id, stat in sorted(
+            first_token_label_dic.items(), key=lambda x: -x[1]["total"]
+        )[:5]:
+            token = tokenizer.decode([token_id])
+            metrics[f"accuracy_first_token_{token}"] = stat["correct"] / stat["total"]
+            metrics[f"accuracy_first_token_{token}_total"] = stat["total"]
```

Comment:
Let's use "accuracy recipient token".  to make it more clear.

Reply:
Since "first token" might be ambiguous.

Reply:
Also, adding "all" token accuracy is helpful. For differentiating between function call's function name accuracy and followup accuracy. Let's add that accuracy too. You may need to manually extract token id of "all"

Reply:
Here I added top-5 top 5 function_names that **appear the most in the data**, so "all" is always expected to be in the result: `accuracy_recipient_token_all`. But I will fix this to include: "all" even it is out of top-5